Consolidate Position classes.

paulp opened the pull request:
Having now been burned several times thinking I had a chunk
of positions under my belt only to find my pants had vanished
when I forged ahead, I am carving out some intermediate stages
for positions.

This one only reshuffles the internals. It consolidates almost
all the behavior into one class, leaving little stub subclasses
to mimic the classes which came before - since nothing is final
or has access anything less than public, there's no way to touch
any constructor without breakage, so they all have to stay.

I removed redundant/incorrect documentation and rewrote the
rest of it, although I don't expect the lifespan of any of it
to be enormous.

The overall thinking behind the existing design was a bit elusive.
Concrete exception-throwing implementations were provided in the base
class for every method, with less exceptional overrides defined
in selected subclasses. If you're going to define every method
in the base class, concretely no less, then there is little
advantage and lots of disadvantage to spreading out over a bunch
of subclasses.

Note that now Position contains no state and has no constructor,
characteristics which behoove its possibly distant ambition
toward AnyValhood. Though that does ignore the "api" class it
inherits which brims with implementation detail, api.Position.
It is a burden which will likely prove heavy.

paulp committed:
Consolidate Position classes.

Having now been burned several times thinking I had a chunk
of positions under my belt only to find my pants had vanished
when I forged ahead, I am carving out some intermediate stages
for positions.

This one only reshuffles the internals. It consolidates almost
all the behavior into one class, leaving little stub subclasses
to mimic the classes which came before - since nothing is final
or has access anything less than public, there's no way to touch
any constructor without breakage, so they all have to stay.

I removed redundant/incorrect documentation and rewrote the
rest of it, although I don't expect the lifespan of any of it
to be enormous.

The overall thinking behind the existing design was a bit elusive.
Concrete exception-throwing implementations were provided in the base
class for every method, with less exceptional overrides defined
in selected subclasses. If you're going to define every method
in the base class, concretely no less, then there is little
advantage and lots of disadvantage to spreading out over a bunch
of subclasses.

Note that now Position contains no state and has no constructor,
characteristics which behoove its possibly distant ambition
toward AnyValhood. Though that does ignore the "api" class it
inherits which brims with implementation detail, api.Position.
It is a burden which will likely prove heavy.

paulp commented:
Review by @retronym. I'd love to get this one in sooner than later, as it will assist me in making a new base camp further up Everest.

dragos commented:
@paulp thanks a lot for pushing for this! I'll look at the PR tomorrow, in the meantime please note this breaks one of our dependencies (scala-refactoring). I'm not saying that's a show stopper, but we probably need to fix that breakage before going further.

```
[ERROR] /localhome/jenkins/b/workspace/pr-scala-integrate-ide/scala-refactoring/org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/transformation/TreeFactory.scala:17: error: illegal inheritance from sealed class Position
[ERROR]   object Invisible extends Position
[ERROR]                            ^
```

paulp commented:
No problem. I was hoping nobody extended Position directly, but it doesn't really matter. I'll unseal it.

paulp commented:
Nobody should be confronted with 8324766 bytes to find out whether a build failed -- this is at least three orders of magnitude too high -- but based on very little, I think it passed.

retronym commented:
Just to be sure, I checked that we didn't have redundant fields between sub/super classes. All clear:

```
scala> :javap -private scala.reflect.internal.util.RangePosition
Compiled from "Position.scala"
public class scala.reflect.internal.util.RangePosition extends scala.reflect.internal.util.OffsetPosition{
    private final int startIn;
    private final int endIn;
    public boolean isRange();
    public int start();
    public int end();
    public scala.reflect.internal.util.RangePosition(scala.reflect.internal.util.SourceFile, int, int, int);
}


scala> :javap -private scala.reflect.internal.util.OffsetPosition
Compiled from "Position.scala"
public class scala.reflect.internal.util.OffsetPosition extends scala.reflect.internal.util.DefinedPosition{
    private final scala.reflect.internal.util.SourceFile sourceIn;
    private final int pointIn;
    public boolean isRange();
    public scala.reflect.internal.util.SourceFile source();
    public int point();
    public int start();
    public int end();
    public scala.reflect.internal.util.OffsetPosition(scala.reflect.internal.util.SourceFile, int);
}


scala> :javap -private scala.reflect.internal.util.TransparentPosition
Compiled from "Position.scala"
public class scala.reflect.internal.util.TransparentPosition extends scala.reflect.internal.util.RangePosition{
    public boolean isTransparent();
    public scala.reflect.internal.util.TransparentPosition(scala.reflect.internal.util.SourceFile, int, int, int);
}
```

retronym commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -1,293 +1,251 @@
+  def includes(pos: Position): Boolean         = isRange && pos.isDefined && start <= pos.startOrPoint && pos.endOrPoint <= end
+  def properlyIncludes(pos: Position): Boolean = includes(pos) && (start < pos.startOrPoint || pos.endOrPoint < end)
+  def precedes(pos: Position): Boolean         = bothDefined(pos) && endOrPoint <= pos.startOrPoint
+  def properlyPrecedes(pos: Position): Boolean = bothDefined(pos) && endOrPoint < pos.startOrPoint
```
8 calls the the newly deprecated `endOrPoint` and `startOrPoint` should be replaced here.

retronym commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -1,293 +1,251 @@
-  def inUltimateSource(source : SourceFile): Position =
-    if (source == null) this else source.positionInUltimateSource(this)
+  def offset(source: SourceFile, point: Int): Position                            = validate(new OffsetPosition(source, point))
+  def range(source: SourceFile, start: Int, point: Int, end: Int): Position       = validate(new RangePosition(source, start, point, end))
```
Presumably we want to route all construction through here eventually. We can start by doing that:

```
% ack -l 'new (Range|Offset|Transparent)Position'
src/compiler/scala/tools/nsc/ast/parser/Parsers.scala
src/compiler/scala/tools/nsc/javac/JavaParsers.scala
src/compiler/scala/tools/nsc/javac/JavaScanners.scala
src/compiler/scala/tools/reflect/MacroImplementations.scala
src/interactive/scala/tools/nsc/interactive/Picklers.scala
src/interactive/scala/tools/nsc/interactive/tests/Tester.scala
src/reflect/scala/reflect/internal/Positions.scala
src/reflect/scala/reflect/internal/util/Position.scala
src/reflect/scala/reflect/internal/util/SourceFile.scala
src/scaladoc/scala/tools/nsc/doc/ScaladocAnalyzer.scala
test/disabled/presentation/simple-tests/src/Tester.scala
test/junit/scala/tools/nsc/symtab/CannotHaveAttrsTest.scala
```

paulp replied:
I will be touching all the call sites in a commit to follow. As I said I am only trying to book some gain with this commit because there are so many moving parts trying to improve positions in a compatible way.

retronym commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -1,293 +1,251 @@
-  def column: Int = throw new UnsupportedOperationException("Position.column")
+  private def validate[T <: Position](pos: T): T = {
+    if (pos.isRange)
+      assert(pos.start <= pos.end, s"bad position: ${pos.show}")
```
Perhaps I'm being pedantic, but I'd prefer the new validation to be added in a separate commit from the reshuffling. Ditto for other new code this commit that I might not have seen.

paulp replied:
That's not new validation. It was line 12 of RangePosition.

paulp replied:
And it's out here because I didn't want any constructor code in Position.

retronym commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -1,293 +1,251 @@
-   */
-  def sameRange(pos: Position): Boolean =
-    isRange && pos.isRange && start == pos.start && end == pos.end
+sealed trait SealedPosition {
```
Should comment to explain the evolutionary motivation for sealed-ness and the name "SealedXxx". At first glance, it is hard to distinguish from other names domain of positions ("Transparent" / "Opaque").

paulp replied:
Well, I was really after "private", not sealed. I'm just trying to minimize the exposed surface area of implementation contributing to the existing subclasses, all of which I hope to deprecate.

retronym commented:
LGTM. thanks for the explanation.

paulp commented:
@dragos When you get a chance would you mind giving this an LGTM or some feedback? I'm off the typesafe clock in about 24 hours. I'll be staying with this until it's done, but I do hope to limit the duration of my volunteerism.

paulp committed:
Removed deprecated calls.

And changed the implementation internal traits to private[util]
rather than sealed because that's closer to the point. I would
make them "private" except that this incurs "private type escapes
scope" errors where private[util] does not.

Also improved file organization.

paulp commented:
Rebased, and added a second commit which removes the deprecated calls and changes "sealed" to "private[util]" where appropriate. I also reorganized the file to put the internal classes at the back of the classroom, which may make the change look larger than it is.

paulp commented:
Okay, circumstances force my hand. Feedback still welcome. There will be more interesting code to feed back on in any case.

som-snytt commented:
I'm glad you have consolidated your position.
